sw/device/mask_rom/mask_rom_start.S

```diff
@@ -0,0 +1,283 @@
+   *
+   * We actually booted with this value of `mtvec`, so we shouldn't need to do
+   * this.
+   *
+   * The lowest two bits should be `0b01` to ensure we use vectored interrupts.
+   */
+  la   t0, _mask_rom_interrupt_vector
+  andi t0, t0, -4
+  ori  t0, t0, 0b01
+  csrw mtvec, t0
+
+  /**
+   * Clean Device State Part 1
+   */
+
+  // Zero all writable registers except x2 (sp).
+  mv x1,  zero
+  // NOT x2 (sp) - We have already set it to the right value above.
+  mv x3,  zero
+  mv x4,  zero
+  mv x5,  zero
+  mv x6,  zero
+  mv x7,  zero
+  mv x8,  zero
+  mv x9,  zero
+  mv x10, zero
+  mv x11, zero
+  mv x12, zero
+  mv x13, zero
+  mv x14, zero
+  mv x15, zero
+  mv x16, zero
+  mv x17, zero
+  mv x18, zero
+  mv x19, zero
+  mv x20, zero
+  mv x21, zero
+  mv x22, zero
+  mv x23, zero
+  mv x24, zero
+  mv x25, zero
+  mv x26, zero
+  mv x27, zero
+  mv x28, zero
+  mv x29, zero
+  mv x30, zero
+  mv x31, zero
+
+  // TODO: Setup SRAM Scrambling
+  // Temporarily: Zero out ram_main
+  //
+  // `t0` is the address to start zeroing at.
+  // `t1` is the address to stop zeroing at.
+  li  t0, TOP_EARLGREY_RAM_MAIN_BASE_ADDR
+  li  t1, (TOP_EARLGREY_RAM_MAIN_BASE_ADDR + TOP_EARLGREY_RAM_MAIN_SIZE_BYTES)
+  bge t0, t1, sram_zero_end
+sram_zero_loop:
+  // TODO: Unroll loop
+  sw   zero, 0x0(t0)
+  addi t0, t0, 0x4
+  blt  t0, t1, sram_zero_loop
+sram_zero_end:
+
+  /**
+   * Setup C Runtime
+   */
+
+  // Initialize the `.data` section.
+  //
+  // `t0` is the start pointer of `.data` (in RAM)
+  // `t1` is the end pointer of `.data` (in RAM)
+  // `t2` is the start pointer of `.data` (in ROM).
+  // `t3` is a scratch register for the copy.
+  la  t0, _data_start
+  la  t1, _data_end
+  la  t2, _data_init_start
+  bge t0, t1, data_copy_loop_end
+data_copy_loop:
+  // TODO: Unroll this loop
+  lw   t3, 0(t2)
+  sw   t3, 0(t0)
+  addi t0, t0, 0x4
+  addi t2, t2, 0x4
+  blt  t0, t1, data_copy_loop
+data_copy_loop_end:
+
+  // Initialize the `.bss` section.
+  //
+  // We do this despite zeroing all of SRAM above, so that we still zero `.bss`
+  // once we've enabled SRAM scrambling.
+  //
+  // `t0` is the address to start zeroing at.
+  // `t1` is the address to stop zeroing at.
+  la  t0, _bss_start
+  la  t1, _bss_end
+  bge t0, t1, bss_zero_end
+bss_zero_loop:
+  // TODO: Unroll loop
+  sw   zero, 0x0(t0)
+  addi t0, t0, 0x4
+  blt  t0, t1, bss_zero_loop
+bss_zero_end:
+
+  // Re-clobber all of the registers used to Setup the C Runtime.
+  mv t0, zero
+  mv t1, zero
+  mv t2, zero
+  mv t3, zero
+
+  // Setup global pointer.
+  //
+  // This requires that we disable linker relaxations, or it will be relaxed to
+  // `mv gp, gp`, so we disabled relaxations for all of `_mask_rom_start_boot`.
+  la gp, __global_pointer$
+
+  /**
+   * Jump to C Code
+   */
+  .extern mask_rom_boot
+  call mask_rom_boot
+
+  // Loop forever if mask_rom_boot somehow returns.
+1:
+  wfi
+  j 1b
+
+  .size _mask_rom_start_boot, .-_mask_rom_start_boot
```

Comment:
Is this for the linker or for debug purposes? Do you mind adding a comment here?

Reply:
It's for both - ELF symbols have a "size" field which tells you how much space that symbol takes up. Labels don't use it, but globals and functions do. I think debuggers use it to help you disassemble functions.